### ISSUE TITLE
Exception on simple put request

### ISSUE THREAD

**Issue by ghristov** [jurmous/etcd4j#26]
Sorry for posting this here but i have the following problem.
I create single get request.

```java
   try{

      EtcdClient client = new EtcdClient(URI.create("http://x.x.x.x:4012"));
      EtcdKeysResponse response = client.put("foo", "bar").send().get();

      // Prints out: bar
      System.out.println(response.node.value);
    }catch(EtcdException e){
      // Do something with the exception returned by etcd
    }catch(IOException e){
      // Exception happened in the retrieval. Do something with it.
    }catch(TimeoutException e){
      // Timeout happened. Do something
    }
```

and there is a thrown exception

> WARNING: An exception was thrown by ousio.etcd4j.transport.EtcdNettyClient$3.operationComplete()
 java.io.IOException:io.netty.handler.codec.http.multipart.HttpPostRequestEncoder$ErrorDataEncoderException: Cannot create a Encoder if not a POST
>	at mousio.etcd4j.transport.EtcdNettyClient.createHttpRequest(EtcdNettyClient.java:276)
>	at mousio.etcd4j.transport.EtcdNettyClient$3.operationComplete(EtcdNettyClient.java:198)
>	at mousio.etcd4j.transport.EtcdNettyClient$3.operationComplete(EtcdNettyClient.java:159)
>	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:680)
>	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:603)
>	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:563)
>	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:406)
>	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
>	at >io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:247)
>	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:282)
>	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
	at java.lang.Thread.run(Thread.java:744)

**Comment by ghristov** [comment]
As far as I can see the constructor of the HttpPostRequestEncoder.java doesn't allow creation of the object  if the method is different from POST

```java
public HttpPostRequestEncoder(
            HttpDataFactory factory, HttpRequest request, boolean multipart, Charset charset,
            EncoderMode encoderMode)
            throws ErrorDataEncoderException {
        if (factory == null) {
            throw new NullPointerException("factory");
        }
        if (request == null) {
            throw new NullPointerException("request");
        }
        if (charset == null) {
            throw new NullPointerException("charset");
        }
        if (request.getMethod() != HttpMethod.POST) {
            throw new ErrorDataEncoderException("Cannot create a Encoder if not a POST");
        }
        
         .......
    }
```

As you can see in the implementation  of setRequestParameters in the EtcdNettyClient you allow creation of this encoder even if the method is PUT. 

```java
 private static HttpRequest setRequestParameters(String uri, EtcdRequest<?> etcdRequest, HttpRequest httpRequest) throws Exception {
    // Set possible key value pairs
    Map<String, String> keyValuePairs = etcdRequest.getRequestParams();
    if (keyValuePairs != null && !keyValuePairs.isEmpty()) {
      HttpMethod etcdRequestMethod = etcdRequest.getMethod();
      if (etcdRequestMethod == HttpMethod.POST || etcdRequestMethod == HttpMethod.PUT) {
        HttpPostRequestEncoder bodyRequestEncoder = new HttpPostRequestEncoder(httpRequest, false);
       
       ....
  }
```

I think this should be fixed.

**Comment by slintes** [comment]
Hi,

I don't see the exception, and I see a different source for HttpPostRequestEncoder, where PUT is allowed:

    public HttpPostRequestEncoder(
            HttpDataFactory factory, HttpRequest request, boolean multipart, Charset charset,
            EncoderMode encoderMode)
            throws ErrorDataEncoderException {
        if (factory == null) {
            throw new NullPointerException("factory");
        }
        if (request == null) {
            throw new NullPointerException("request");
        }
        if (charset == null) {
            throw new NullPointerException("charset");
        }
        HttpMethod method = request.method();
        if (!(method.equals(HttpMethod.POST) || method.equals(HttpMethod.PUT)
                || method.equals(HttpMethod.PATCH) || method.equals(HttpMethod.OPTIONS))) {
            throw new ErrorDataEncoderException("Cannot create a Encoder if not a POST");
        }
...

Do have another netty-codec-http version on your classpath? The one used in etcd4j is io.netty:netty-codec-http:4.1.0.Beta3, see build.gradle

Regards, Marc

**Comment by ghristov** [comment]
Yeah I'm using 4.0.25 Final... And obviously in the newer version they changed the api. I changed it to 4.1.3Beta and it worked. Well i'm not using gradle. I'm using this library from Maven.

**Comment by slintes** [comment]
:thumbsup:

**Comment by ghristov** [comment]
Btw I think project must be packaged with all of it's dependencies because as you can see I can't know which version of the netty api is used. And since i use always final versions that was a problem in this case.

**Comment by slintes** [comment]
I don't agree with this. You can see the dependencies with version numbers in the build.gradle file: https://github.com/jurmous/etcd4j/blob/master/build.gradle#L42. If if you don't use gradle, you still can use that information to add the dependencies to your maven pom file (like you did obviously) or on any other way you prefer to your classpath.
If every library comes packaged with every dependency, you will have chaos on your classpath sooner or later, when different libraries use different versions of dependencies.

Oh wait, it's even easier: why do you not just add etcd4j to your pom file? That should resolve all dependencies automatically, not? http://search.maven.org/#artifactdetails%7Corg.mousio%7Cetcd4j%7C2.1.0%7Cjar

**Comment by slintes** [comment]
Oh, no, won't work. The pom of etcd4j does not contain all dependencies.
@jurmous I think the tooling for uploading to the maven repository does not understand the "clientCompile" you use in the build.gradle file... :confused:

**Comment by jurmous** [comment]
Oh dear. Something is indeed wrong with the gradle build. Gradle maven plugin by default only includes the dependencies of the main sourceSet which in this case does not include Netty. I have opened #27 with a more specific description and will look into it tomorrow. Thanks for reporting this issue in! 

I will also look into back porting to the final Netty 4.0 release since those betas take much longer time to mature than I was expecting when I started this project.

**Comment by ghristov** [comment]
Glad that i helped :)

**Comment by jurmous** [comment]
Released in 2.1.1 :)  

I could not easily convert back to Final because of the issue reported above which prevented to use the Post decoder with the Put. So I left the dependency at Beta for this moment. Maybe I will look into this issue later if I can solve it differently.

**Comment by ghristov** [comment]
Now it works like a charm. No need to add any additional libraries.

Just added this one and all deps are resolved and included.

<dependency>
      <groupId>org.mousio</groupId>
      <artifactId>etcd4j</artifactId>
      <version>2.1.1</version>
    </dependency>
